### PR TITLE
[SPARK-40534][CONNECT][FOLLOW-UP] Support all join types in python client

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Following up https://github.com/apache/spark/pull/38157, add all join types support for python. Please note that in the PR we decided to not support CROSS Join now.

### Why are the changes needed?

Finish the join type support in Connect. 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @zhengruifeng @HyukjinKwon

File: python/pyspark/sql/connect/plan.py [hunk 1]
```diff
@@ -368,21 +368,45 @@ def __init__(
...
+            join_type = proto.Join.JoinType.JOIN_TYPE_FULL_OUTER
+        elif how in ["leftouter", "left"]:
+            join_type = proto.Join.JoinType.JOIN_TYPE_LEFT_OUTER
+        elif how in ["rightouter", "right"]:
+            join_type = proto.Join.JoinType.JOIN_TYPE_RIGHT_OUTER
+        elif how in ["leftsemi", "semi"]:
+            join_type = proto.Join.JoinType.JOIN_TYPE_LEFT_SEMI
+        elif how in ["leftanti", "anti"]:
+            join_type = proto.Join.JoinType.JOIN_TYPE_LEFT_ANTI
+        else:
```
Comment (amaliujia) [inline python/pyspark/sql/connect/plan.py:391]: Note that in https://github.com/apache/spark/pull/38157 the feedback was to not support CROSS join so it is not listed here. Whenever it makes sense to have CROSS join then adding the support will be easy and fast.

**Review by HyukjinKwon — APPROVED** [review]

**Review by zhengruifeng — APPROVED** [review]
LGTM

File: python/pyspark/sql/connect/plan.py (hunk 1 above)
Comment (hvanhovell) [reply python/pyspark/sql/connect/plan.py:391]: Why is that? Please keep in mind that these type of decisions increase the pain of porting a pyspark job over to connect.

**Review by hvanhovell — APPROVED** [review]
LGTM

File: python/pyspark/sql/connect/plan.py (hunk 1 above)
Comment (amaliujia) [reply python/pyspark/sql/connect/plan.py:391]: There were some different opinions for why that is useful. If you think we probably need that back, let's have a discussion first to reach a consensus.

**Comment by amaliujia** [comment]
@HyukjinKwon This PR should be ready to merge.

**Comment by zhengruifeng** [comment]
Merged into master

**Pull request closed without merging**